otc-swap-predicate/tests/harness.rs

```diff
@@ -0,0 +1,145 @@
+    let predicate_coin_utxo_id = predicate_coin.utxo_id.clone().into();
+
+    // Get other coin to spend
+    let swap_coin = &provider.get_coins(&wallet.address()).await.unwrap()[0];
+    let swap_coin_utxo_id = swap_coin.utxo_id.clone().into();
+    let swap_coin_amount: u64 = swap_coin.amount.clone().into();
+
+    // Configure inputs and outputs to send coins from the predicate root to another address
+
+    // The predicate allows to spend its tokens if `ask_amount` is sent to the offer maker.
+    // This must match the amount in the predicate
+    let ask_amount = 42;
+
+    // Coin belonging to the predicate root
+    let input_predicate = Input::CoinPredicate {
+        utxo_id: predicate_coin_utxo_id,
+        owner: predicate_root,
+        amount: offered_amount,
+        asset_id: base_asset,
+        maturity: 0,
+        predicate: predicate_bytecode,
+        predicate_data: vec![1u8, 0u8], // Predicate data is the index of the input and output that pay the receiver
+    };
+
+    // Coin belonging to the wallet taking the order
+    let input_from_taker = Input::CoinSigned {
+        utxo_id: swap_coin_utxo_id,
+        owner: wallet.address(),
+        amount: swap_coin_amount,
+        asset_id: base_asset,
+        witness_index: 0,
+        maturity: 0,
+    };
+
+    // Output for the coin transferred to the receiver
+    let output_to_receiver = Output::Coin {
+        to: receiver_address,
+        amount: ask_amount,
+        asset_id: base_asset,
+    };
+
+    // Output for the coin transferred to the order taker
+    let output_to_taker = Output::Coin {
+        to: wallet.address(),
+        amount: offered_amount,
+        asset_id: base_asset,
+    };
+
+    // Change output for unspent fees
+    let output_change = Output::Change {
+        to: wallet.address(),
+        amount: 0,
+        asset_id: Default::default(),
+    };
+
+    let mut tx = Transaction::Script {
+        gas_price: 0,
+        gas_limit: 10_000_000,
+        maturity: 0,
+        byte_price: 0,
+        receipts_root: Default::default(),
+        script: Opcode::RET(REG_ONE).to_bytes().to_vec(),
+        script_data: vec![],
+        inputs: vec![input_predicate, input_from_taker],
+        outputs: vec![output_to_receiver, output_to_taker, output_change],
+        witnesses: vec![],
+        metadata: None,
+    };
+
+    // Sign and execute the transaction
+    wallet.sign_transaction(&mut tx).await.unwrap();
+    let script = Script::new(tx);
+    let _receipts = script.call(&client).await.unwrap();
+
+    let predicate_balance = utils::get_balance(&provider, predicate_root, base_asset).await;
+    let wallet_balance = utils::get_balance(&provider, wallet.address(), base_asset).await;
+    let receiver_balance = utils::get_balance(&provider, receiver_address, base_asset).await;
+
+    // The predicate root's coin has been spent
+    assert_eq!(predicate_balance, 0);
```

Comment:
It would be nice to see a negative test or two, that the coin isn't spendable if the conditions are violated.

Reply:
hmm is there currently a way to test this? Since the transaction isn't reverting, but rather invalid :

`ContractCallError("Response errors; Invalid transaction: The transaction contains a predicate which failed to validate`